space2stats_api/src/space2stats/h3_utils.py

```diff
@@ -1,145 +1,79 @@
 
-    if isinstance(aoi_shape, MultiPolygon):
-        h3_ids = set()
-        for geom in aoi_shape.geoms:
-            # Convert each geometry to GeoJSON
-            geom_geojson = mapping(geom)
-            h3_ids.update(
-                _generate_h3_ids(geom_geojson, resolution, spatial_join_method)
-            )
-    else:
-        # Single polygon case
-        h3_ids = _generate_h3_ids(aoi_geojson, resolution, spatial_join_method)
+    # Generate H3 IDs as uint64
+    h3_ids_uint64 = geometry_to_cells(
+        geom, resolution, containment_mode=containment_mode
+    )
 
-    return h3_ids
+    return h3_ids_uint64
 
 
 def generate_h3_geometries(
-    h3_ids: List[str], geometry_type: Optional[str]
-) -> List[Dict[str, Any]]:
+    h3_ids_uint64: List[int], geometry_type: Literal["polygon", "point"] = "polygon"
+) -> List[Dict]:
     """
-    Generate geometries (polygon or point) for a list of H3 hexagon IDs.
+    Generate geometries (GeoJSON polygons or points) for a list of H3 hexagon IDs (as uint64).
 
     Parameters:
-        h3_ids (List[str]): A list of H3 hexagon IDs.
+        h3_ids_uint64 (List[int]): A list of H3 hexagon IDs in uint64 format.
         geometry_type (Optional[str]):
             The type of geometry to generate ('polygon' or 'point').
 
     Returns:
-        List[Dict[str, Any]]: A list of geometries in GeoJSON format.
+        List[Dict]: A list of GeoJSON geometries (Polygon or Point).
     """
-    if geometry_type not in ["polygon", "point"]:
-        raise ValueError("Invalid geometry type")
-
-    geometries = []
-    for h3_id in h3_ids:
-        if geometry_type == "polygon":
-            hex_boundary = h3.h3_to_geo_boundary(h3_id, geo_json=True)
-            geometries.append(mapping(Polygon(hex_boundary)))
-        elif geometry_type == "point":
-            # h3_to_geo does not have geo_json parameter to invert order of coords
-            x, y = h3.h3_to_geo(h3_id)
-            geometries.append(mapping(Point(y, x)))
+    if geometry_type == "polygon":
+        # Convert H3 cells to WKB polygons
+        wkb_geometries = cells_to_wkb_polygons(h3_ids_uint64)
+    elif geometry_type == "point":
+        # Convert H3 cells to WKB points (centroid of each hexagon)
+        wkb_geometries = cells_to_wkb_points(h3_ids_uint64)
+    else:
+        raise ValueError("Invalid geometry type. Use 'polygon' or 'point'.")
+
+    # Convert pyarrow Array to GeoJSON geometries
+    geojson_geometries = []
+    for geom in wkb_geometries:
+        if isinstance(geom, pa.LargeBinaryScalar):
+            # Convert the LargeBinaryScalar to bytes
+            shapely_geom = wkb.loads(geom.as_py())
+            # Convert Shapely geometry to GeoJSON-like dict
+            geojson_geometries.append(mapping(shapely_geom))
+        else:
+            raise TypeError(
+                f"Expected pa.LargeBinaryScalar, but got {type(geom).__name__}"
+            )
```

Comment:
This loop will be really slow. You can instead use `shapely.from_wkb` right on the pyarrow array of WKB geometries.

```
shapely.to_geojson(shapely.from_wkb(wkb_geometries))
```

Reply:
ah, good catch. I'll have to update the API structure and there's a little mismatch while still use the coordinate approach. Might be worth tackling in a separate PR? Performance has improved tremedously already.

Reply:
just implemented.